compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -0,0 +1,158 @@
+                // always inserts a move from that argument to a fresh temporary and then passes
+                // *that* temporary to the call, so this behavior of MIR is never actually
+                // observed. But we ought to account for it in order for this optimization to
+                // respect the semantics of MIR as they are today.
+                //
+                // We therefore treat moves into a call argument as mutating uses. This should
+                // never actually happen for arguments in practice, but let's cover all our bases.
+                //
+                // Another way to look at this is that we're treating moves as mutations, as they
+                // technically are in MIR, and replacing moves with copies everywhere but in
+                // function arguments. This works out to the same thing.
+
+                self.visit_operand(func, location);
+
+                for arg in args {
+                    match *arg {
+                        Operand::Move(ref place) => {
+                            // This would ordinarily be
+                            // `PlaceContext::NonMutatingUse(NonMutatingUseContext::Move)`, but we
+                            // need a mutating use. Just pretend it's a store.
+                            self.visit_place(
+                                place,
+                                PlaceContext::MutatingUse(MutatingUseContext::Store),
+                                location,
+                            );
+                        }
+                        _ => self.super_operand(arg, location),
+                    }
+                }
+
+                self.visit_place(
+                    destination,
+                    PlaceContext::MutatingUse(MutatingUseContext::Call),
+                    location,
+                );
+            }
+            _ => self.super_terminator(terminator, location),
+        }
+    }
+}
+
+/// Returns the deduced parameter attributes for a function.
+///
+/// Deduced parameter attributes are those that can only be soundly determined by examining the
+/// body of the function instead of just the signature. These can be useful for optimization
+/// purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+/// dependent crates can use them.
+pub fn deduced_param_attrs<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx [DeducedParamAttrs] {
+    // For now, don't deduce any parameter attributes on intrinsics.
+    //
+    // FIXME(pcwalton): Deduce some?
+    if tcx.is_intrinsic(def_id) {
+        return &[];
+    }
+
+    // Don't deduce any parameter attributes for trait methods without default implementations,
+    // since they have no MIR.
+    if tcx.trait_of_item(def_id).is_some() {
+        return &[];
+    }
+
+    // Don't deduce any parameter attributes on foreign items, since they also have no MIR.
+    if tcx.is_foreign_item(def_id) {
```

Comment:
Would testing `!tcx.is_mir_available(def_id)` be more accurate?

Reply:
😆 I was too slow with my comment